### ISSUE TITLE
Join forces?

### ISSUE THREAD

**Issue by qwertie** [ZacharyPatten/Towel#45]
I notice you're implementing a set of functionality that is very similar to the kinds of things I've implemented in my own libraries - collection stuff including several data structures, geometry, generic math, a few sort functions...

Even `Symbolics.Expression` seems vaguely related to my work on a [universal syntax tree](http://loyc.net/loyc-trees) and a [standard expression language](http://loyc.net/les).

Plus, 15 years ago I implemented units-of-measure-inference support as an add-on to an obscure .NET programming language called boo, and I just haven't had time to do something like that for C# (I am of the opinion that unit checking works better as a built-in language feature rather than as a library, but language features are more difficult to do.)

What would you think about joining forces and merging our libraries? Here's the [web site for mine](http://core.loyc.net/).

**Comment by qwertie** [comment]
P.S. I arrived here via your article "Accessing XML Documentation via Reflection" - but I can't find any code related to XML documentation here, at least not via GitHub search.

**Comment by ZacharyPatten** [comment]
I'm definitely open to suggestions (especially areas/topics to branch out to).

If you are interested in contributing to Towel, then you can open issues for individual components you would like to contribute/sugguest (and we could discuss each individually).

If you want to pull in portions of Towel into your project, then I will likely be okay with that as well as long as you provide some citations as to the original author(s) of the portions you copy. You can DM me on Discord to discuss anything.

Also, yes the code related to the MSDN article is currently in the `Towel.Meta` class. There are examples here:
https://github.com/ZacharyPatten/Towel/blob/860f46bc5fd359967d7c3c265e87dab444c3015e/Examples/BasicsAndExtensions/Program.cs#L198
And unit tests here:
https://github.com/ZacharyPatten/Towel/blob/master/Tools/Towel_Testing/Meta.cs

**Comment by qwertie** [comment]
I'm not proposing that I simply make pull requests, or copy code to my project. I'm proposing merging our projects into a single project. So... if that sounds interesting to you, have a look around my web sites (there are three - http://core.loyc.net, http://loyc.net and http://ecsharp.net) and give it some thought.

**Comment by qwertie** [comment]
Actually, two projects would probably be best, with the stuff on the third web site (EC#/LeMP/LLLPG) split out in a second repo. It used to be split that way originally - I was using `git subtree` to put the core repo inside the main one, but a few years back `git subtree` stopped working and I didn't care enough to track down the cause.

I'm open to changing the branding of my own code to Towel. This is because the concept of Loyc is not intended to be .NET specific, but the current libraries are overly .NET specific. A rename could pave the way to later introduce new Loyc libraries that are not .NET-specific.

**Comment by qwertie** [comment]
I guess this means you are not interested... may I ask why?

**Comment by ZacharyPatten** [comment]
Sorry... I am actually just way too busy right now. There are so many things on my backlog, and I've been working on other projects too (including [dotnet-console-games](https://github.com/ZacharyPatten/dotnet-console-games). The little bit of time I have had to explore your projects gives me the initial impression that although they cover the same concepts the code itself it quite different in design and would likely be an extensive process to merge.

I'm not sure when I'll have the time to devote to investigating a potential merge sooner.

**Comment by ZacharyPatten** [comment]
@qwertie I reviewed your code a bit more and here are some further comments I have:

- You are targeting .NET Framework which is now legacy. Towel currently targets .NET Standard 2.0, and I plan on keeping relatively up to date on that. I will probably wait till .NET 5 though to go beyond 2.0 as .NET framework still supports .NET Standard 2.0.
- Your pattern for mathematics (with interfaces) does not scale and is hard to maintain and/or requires excessive code gen. You must make an implementation for every type you want to support, and you do not support custom types.
  - Towel takes a very different approach. It uses runtime compilation via LINQ expressions to wrap the mathematics operations at runtime. This prevents the need to implement algorithms for every type and works with any type that contains the required operators at compile time. For example, I can use the `Length<float>` type inside the `Vector<Length<float>>` type without a custom implementation.
  - There is a proposal for a feature in a future version of C# currently called "Shapes." [Here is a link](https://github.com/dotnet/csharplang/issues/164). If that feature was added to C#, it would remove the need to use interfacing altogether, which would drastically reduce the size of your mathematics code. Even if that feature was added, there are still some benefits to using delegates (which Towel uses) as opposed to shapes, since delegates are redirect-able at runtime.
- I don't use most of the interfaces in the C# libraries (IComparable, IEquateable, etc). This is because I am using generic arguments (along with structs that have implicit converters from delegates) instead. Even in my data structures I am following this pattern. This is a completely different style/approach to how your data structures and algorithms appear to be coded. Why am I using so many generic parameters? Because of the optimizations outlined here: https://github.com/dotnet/csharplang/issues/2904

Due to Towel's heavy usage of runtime compilation and/or generic arguments (as opposed to standard interfacing), there are just so many things that are inconsistent between Towel and the Loyc libraries. Attempting to merge them would inevitable result in huge refactoring of either libraries.
